### ISSUE TITLE
Data loss on simultaneous browsers

### ISSUE THREAD

**Issue by hakebi** [Cleod9/syncmarx-webext#12]
There should be a warning somewhere saying that this extension is not meant to be used at the same time in different computers or you will **lose** data.

The sync works in a way that if the remote has been modified by another computer, it always “wins” an destroy the local changes. I think it only push the local changes if the remote hasn't been modified, otherwise, remote will overwrite the local bookmarks.

Let’s see an example:

We are using our browsers in two computers, in `computer1` we add a new `bookmarkA` to a `folderX`, but we have a default 15min interval sync, so there is no sync for now, It will still take 5 more minutes for it to sync automatically. Now, in `computer2`, we also add a new `bookmarkB` to a the same `folderX`, there is no sync either, in this computer it will still take 8 minutes to automatically sync.

In this scenario, the computer who sync faster is going to "win" and overwrite the changes in the other, causing data loss. Following the example, after 5 min, `computer1` will sync, and the remote will be updated with the new `bookmarkA` in place. But then, 3 min after, the `computer2` will sync, it will check the date of the remote and overwrite the local bookmarks with the remote ones, making the `bookmarkB` disappear forever.

In conclusion, you have to make sure you have synchronized, before making changes in any other browser o data will be lost, so basically don’t make changes in two browsers at the same time.

**Comment by Cleod9** [comment]
Thank you for the report, you are correct that this is an outstanding issue. Bookmarking across different browsers and computers faster than the sync time can result in lost bookmarks, since the most recently synced machine is treated as the source of truth. The extension is labeled as alpha since it still has a few quirks like this. It's a tricky issue since if you are manipulating bookmarks across multiple instances there becomes a certain point where reconciling the order becomes rather complicated.  I hope to address this problem as soon as I have more free time to contribute to the project.

**Comment by hakebi** [comment]
I’ve been looking at the code and thinking of ways of fixing it but as you say, trying to reconcile the order it’s a nightmare, I will think on it and if I get a solution, I will try to implement it.

**Comment by Cleod9** [comment]
Thank you, I'm always up for ideas. It would take some time though, based on how it's implemented I may need to replace the diffing strategy entirely, but don't want to risk breaking it so it may make more sense for me to integrate comprehensive unit testing first

**Comment by Iiridayn** [comment]
Don't strictly need comprehensive unit testing - _any_ unit testing is better than none, especially for critical code.

Generally timestamps are not very reliable in distributed syncing (for example, they are _not_ monotonically increasing, even on the same machine, though it's much worse when multiple machines can all set their own clocks - https://infiniteundo.com/post/25326999628/falsehoods-programmers-believe-about-time has an entertaining list of thing that have gone wrong in the past; ie "Ok, but the time on the server clock and time on the client clock would never be different by a matter of decades.", or "Fine, but the duration of one minute on the system clock would never be more than an hour.
"). Git uses an append-only chain, which is an excellent approach.

A simpler (and weaker) approach I've used in the past was a version counter per tree node, and only updating the lowest node required. When the tree was synced, any subtrees had their versions compared. It would merge anything it could (detecting add/removal; updates required version comparison). I believe only the server was supposed to update the version, though a malicious client could of course have claimed whatever - we weren't concerned about malicious clients with valid credentials (they could do plenty of damage via regular updates/deletion). It had about 1000 lines of tests with simple up to complex cases, and worked okay. Can't remember all the details - was about 3-4 years ago at a former employer.

The only theoretical weakness I recall was deleting a leaf from one computer, and deleting and re-adding an identical leaf from another, with the second (delete/re-add) syncing first - this would delete it a second time, though the deletion intent was just the first. Since that seemed a minor issue, we were pretty happy with it, and moved on. This would also fail to catch copy-delete (if ids are used instead of strings, renames are trivial) with an update inside the former tree, while git's style would detect that as a rename (since that's how git handles renames).

**Comment by Cleod9** [comment]
@Iiridayn Thanks for the tips, I very much agree. I actually filed an issue for unit tests myself on my own repo early last year:

https://github.com/Cleod9/syncmarx-webext/issues/1

I will definitely keep your advice in mind when working on a solution for better syncing, and I'll make sure it's got some kind of unit testing before shipping it (as it would be a breaking change and require data to be migrated safely). It may take some time for me to do this since I've been very busy, but I have not abandoned the project just yet (I still use the extension regularly myself).
